Python/bytecodes.c

```diff
@@ -2266,41 +2264,29 @@ dummy_func(
 
         macro(CONTAINS_OP) = _SPECIALIZE_CONTAINS_OP + _CONTAINS_OP;
 
-        inst(CONTAINS_OP_LIST, (unused/1, left, right -- b)) {
-            DEOPT_IF(!PyList_CheckExact(right));
-            int res = _PyList_Contains(right, left);
-            DECREF_INPUTS();
-            ERROR_IF(res < 0, error);
-            b = (res ^ oparg) ? Py_True : Py_False;
-        }
-
         inst(CONTAINS_OP_SET, (unused/1, left, right -- b)) {
-            DEOPT_IF(!PySet_CheckExact(right));
+            DEOPT_IF(!(PySet_CheckExact(right) || PyFrozenSet_CheckExact(right)));
+            STAT_INC(CONTAINS_OP, hit);
+            // Note: both set and frozenset use the same seq_contains method!
             int res = _PySet_Contains((PySetObject *)right, left);
             DECREF_INPUTS();
             ERROR_IF(res < 0, error);
             b = (res ^ oparg) ? Py_True : Py_False;
         }
 
-        inst(CONTAINS_OP_TUPLE, (unused/1, left, right -- b)) {
-            DEOPT_IF(!PyTuple_CheckExact(right));
-            int res = _PyTuple_Contains((PyTupleObject *)right, left);
-            DECREF_INPUTS();
-            ERROR_IF(res < 0, error);
-            b = (res ^ oparg) ? Py_True : Py_False;
-        }
-
         inst(CONTAINS_OP_DICT, (unused/1, left, right -- b)) {
             DEOPT_IF(!PyDict_CheckExact(right));
+            STAT_INC(CONTAINS_OP, hit);
             int res = PyDict_Contains(right, left);
             DECREF_INPUTS();
             ERROR_IF(res < 0, error);
             b = (res ^ oparg) ? Py_True : Py_False;
         }
 
-        inst(CONTAINS_OP_STR, (unused/1, left, right -- b)) {
-            DEOPT_IF(!PyUnicode_CheckExact(right));
-            int res = PyUnicode_Contains(right, left);
+        inst(CONTAINS_OP_MAPPINGPROXY, (unused/1, left, right -- b)) {
```

Comment:
I'm surprised that this is worthwhile. 
OOI, what are the numbers for this?

Reply:
It might not be. I might just remove it.
On `python.exe -m test_typing test_re test_dis test_zlib.`, I get:
```
frozenset:278
tuple:261
dict:256
list:225
set:150
str:143
mappingproxy:62
dict_values:9
EnumDict:5
collections.OrderedDict:5
collections.defaultdict:4
_Environ:3

```

Reply:
Note: I'm not specializing for tuple or list because it's dominated by searching up through the collection anyways so it's not worth it. The rest are O(1) so maybe worth to cust the constant factor.